src/rubrix/client/datasets.py

```diff
@@ -0,0 +1,780 @@
+#  coding=utf-8
+#  Copyright 2021-present, the Recognai S.L. team.
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+import logging
+from typing import Dict, List, Optional, Tuple, Type, Union
+
+import pandas as pd
+from pkg_resources import parse_version
+
+from rubrix.client.models import (
+    Record,
+    Text2TextRecord,
+    TextClassificationRecord,
+    TokenAttributions,
+    TokenClassificationRecord,
+)
+from rubrix.client.sdk.datasets.models import TaskType
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class DatasetBase:
+    """The Dataset classes are containers for Rubrix records.
+
+    This is the base class to facilitate the implementation for each record type.
+
+    Args:
+        records: A list of Rubrix records.
+
+    Raises:
+        WrongRecordTypeError: When the record type in the provided
+            list does not correspond to the dataset type.
+    """
+
+    _RECORD_TYPE = None
+
+    def __init__(self, records: Optional[List[Record]] = None):
+        if self._RECORD_TYPE is None:
+            raise NotImplementedError(
+                "A Dataset implementation has to define a `_RECORD_TYPE`!"
+            )
+
+        self._records = records or []
+        if self._records:
+            self._validate_record_type()
+
+    def _validate_record_type(self):
+        """Validates the record type.
+
+        Raises:
+            WrongRecordTypeError: When the record type in the provided
+                list does not correspond to the dataset type.
+        """
+        record_types = {type(rec): None for rec in self._records}
+        if len(record_types) > 1:
```

Comment:
Just a suggestion but, for large records list, we could raise the error once we find an unexpected record instead of traverse the whole list

Reply:
Good point! I already have a list of improvements in mind, that I will put in an issue. This PR is already too big and took too long, so I prefer to get this merged and improve the stuff in follow-up PRs.

Reply:
Ok. From my part the only required thing would be to include changes in the tutorial